pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java

```diff
@@ -43,6 +42,12 @@
 
     private final PersistentTopic topic;
 
+    private volatile PositionImpl maxReadPosition = PositionImpl.latest;
+
+    private final LinkedMap<TxnID, PositionImpl> ongoingTxns = new LinkedMap<>();
+
+    private final LinkedMap<TxnID, PositionImpl> aborts = new LinkedMap<>();
```

Comment:
Is there any test coverage? We'd better to add some unit test to make sure the data should be removed correctly to avoid memory lead and make the maxReadPosition is correct.

Reply:
I have add the LowWaterMarkTest. Add the abort and commit MaxReadPosition Test.